### PR TITLE
Heuristics

### PR DESCRIPTION
**Pull request by jiwalker-usgs**
Fairly large refactor, essentially moved everything to a visitor.

### CODE REVIEW

**Comment by jiwalker-usgs** [comment]
Also, sorry about the formatting changes, it is difficult to live in both worlds (tabs vs. spaces) so I've been converting files to tabs as I encounter them.

File: gdp-process-wps/src/main/java/gov/usgs/cida/gdp/wps/algorithm/FeatureWeightedGridStatisticsAlgorithm.java [hunk 1]
```diff
@@ -45,263 +49,224 @@
...
+
+		/*
+		 * Lets run our geometry size heuristic to see if we should go ahead and process
+		 * this bounds.  All geometry size should be relative regardless of the dataset
+		 * variable.
+		 * 
+		 * Lets get the first variable and create a GridDatatype
+		 */
+		if (datasetId.isEmpty()) {
+			log.error("Error subsetting gridded data.  Grid variable list is empty! ");
```
Comment (duselmann) [inline gdp-process-wps/src/main/java/gov/usgs/cida/gdp/wps/algorithm/FeatureWeightedGridStatisticsAlgorithm.java]: I am not a fan of repeating strings.

File: gdp-process-wps/src/main/java/gov/usgs/cida/gdp/wps/algorithm/FeatureWeightedGridStatisticsAlgorithm.java [hunk 2]
```diff
@@ -45,263 +49,224 @@
...
+			addError("Error subsetting gridded data.  Grid variable list is empty! ");
+			return;
```
Comment (duselmann) [inline gdp-process-wps/src/main/java/gov/usgs/cida/gdp/wps/algorithm/FeatureWeightedGridStatisticsAlgorithm.java]: I am not a fan of repeating strings.

**Comment by duselmann** [comment]
I find it extremely difficult to review this with the expansive (and seemingly unnecessary) white space changes. Personally, I prefer the 4 spaces over the tab char but that aside. I cannot trivially tell what changed. I was able to pull the larger sections into files and use a compare too to prove that the few lines that changed are fine. I made it 2/3 the way through this PR. If the subsequent 1/3 is similar to the prior 2/3 then it is okay. Ivan, how are you able to 'see' in order to make those poignant comments? Do you have a better compare tool than the web page?

File: gdp-process-wps/src/test/java/gov/usgs/cida/gdp/wps/algorithm/heuristic/ResultSizeAlgorithmHeuristicTest.java [hunk 3]
```diff
@@ -32,229 +32,233 @@
...
+		}
+
+		coloradoFeatureCollection = FileDataStoreFinder.getDataStore(
+				ResultSizeAlgorithmHeuristicTest.class.getClassLoader().getResource("shp/colorado/CONUS_States.shp"))
+				.getFeatureSource().getFeatures();
+	}
+
+	@AfterClass
+	public static void tearDownAll() {
+		try {
```
Comment (duselmann) [inline gdp-process-wps/src/test/java/gov/usgs/cida/gdp/wps/algorithm/heuristic/ResultSizeAlgorithmHeuristicTest.java]: Why not use the close utility method from commons? The added white space from on the ignore catch {} is not adding any more clarity.

Comment (jiwalker-usgs) [inline gdp-process-wps/src/test/java/gov/usgs/cida/gdp/wps/algorithm/heuristic/ResultSizeAlgorithmHeuristicTest.java]: Unfortunately the GridDataset doesn't implement any interface that makes it easy to close (Closeable, AutoCloseable, etc) so we are stuck doing this.

**Pull request merged**